esr_key_value_database_intrinsics.md

```diff
@@ -0,0 +1,264 @@
+returned by `kv_it_create` or was destroyed.
+
+| Return Value | Description |
+|--------------|-------------|
+| -1           | `itr`'s key is less than `key` |
+| 0            | `itr`'s key is the same as `key` |
+| 1            | `itr`'s key is greater than `key` |
+
+If `itr` has status `iterator_erased`, then `kv_it_key_compare` uses the value of the erased key
+during comparison. If `itr` has status `iterator_oob`, then it compares greater than 
+`key`, no matter what value `key` has.
+
+### kv_it_move_to_oob
+
+```c++
+it_stat  kv_it_move_to_oob(uint32_t itr);
+```
+
+Move `itr` to out-of-bounds and return the new status (`iterator_oob`). It aborts the
+transaction if `itr` wasn't returned by `kv_it_create` or was destroyed.
+
+### kv_it_increment
+
+```c++
+it_stat  kv_it_increment(uint32_t itr);
+```
+
+Increment an iterator and return its new status. It aborts the
+transaction if `itr` wasn't returned by `kv_it_create` or was destroyed.
+
+If `itr`'s status is `iterator_ok` or `iterator_erased`, then this finds the
+next non-deleted key in range. If found, the new status is `iterator_ok`. If
+not found, the new status is `iterator_oob`.
+
+If `itr`'s status is `iterator_oob` then this finds the first non-deleted key
+in range. If found, the new status is `iterator_ok`. If not found, the new
+status is `iterator_oob`.
+
+`kv_it_increment` never returns `iterator_erased`.
+
+### kv_it_decrement
+
+```c++
+it_stat  kv_it_decrement(uint32_t itr);
+```
+
+Decrement an iterator and return its new status. It aborts the
+transaction if `itr` wasn't returned by `kv_it_create` or was destroyed.
+
+If `itr`'s status is `iterator_ok` or `iterator_erased`, then this finds the
+previous non-deleted key in range. If found, the new status is `iterator_ok`. If
+not found, the new status is `iterator_oob`.
+
+If `itr`'s status is `iterator_oob` then this finds the last non-deleted key
+in range. If found, the new status is `iterator_ok`. If not found, the new
+status is `iterator_oob`.
+
+`kv_it_decrement` never returns `iterator_erased`.
+
+### kv_it_lower_bound
+
+```c++
+it_stat  kv_it_lower_bound(uint32_t itr, const char* key, uint32_t size);
+```
+
+Find the first non-deleted key >= the provided key and return the new iterator status.
+It aborts the transaction if `itr` wasn't returned by `kv_it_create` or was destroyed.
+
+If a key is found, the new status is `iterator_ok`. If not found, the new
+status is `iterator_oob`. `kv_it_lower_bound` never returns `iterator_erased`.
+
+### kv_it_key
+
+```c++
+it_stat  kv_it_key(uint32_t itr, uint32_t offset, char* dest, uint32_t size, uint32_t& copied);
+```
+
+Fetch the key from the iterator and return the iterator's status. It aborts the transaction if
+`itr` wasn't returned by `kv_it_create` or was destroyed.
+
+If `size == 0`, then this sets `copied` to the size of the key, but does not fill `dest`. If
+`size > 0`, then this copies up to `size` bytes into `dest` and sets `copied` to the number of
```

Comment:
It should set `copied` to the actual size of the key since that provides more information to the caller. Also, in that case, `copied` should be renamed to `actual_size`.

Reply:
Changed